src/repository/fetcher.ts

```diff
@@ -0,0 +1,37 @@
+import { EventEmitter } from 'events';
+import { ClientFeaturesDelta, ClientFeaturesResponse } from '../feature';
+import { Mode } from '../unleash-config';
+
+export interface FetcherInterface extends EventEmitter {
+  start(): Promise<void>;
+  stop(): void;
+}
+
+export interface FetchingOptions extends PollingFetchingOptions, StreamingFetchingOptions {}
+
+export interface CommonFetchingOptions {
+  url: string;
+  appName: string;
+  instanceId: string;
+  headers?: any;
+  connectionId: string;
+  onSave: (response: ClientFeaturesResponse, fromApi: boolean) => Promise<void>;
+  onSaveDelta: (delta: ClientFeaturesDelta) => Promise<void>;
+  onModeChange?: (mode: 'polling' | 'streaming') => Promise<void>;
```

Comment:
I'm still not sure why the individual fetchers need to know about modes. Shouldn't that be entirely left up to the adaptive fetcher? But it looks like your adding some safeguards down the road, so that's probably a good idea. If this is intentional, I'm sure it's good. Just wanted to double check that it actually is.

Reply:
It is intentional. We can switch modes from the SDK itself by calling setExperimentalMode but also we can switch it from the server driven protocol with either HTTP header for polling or special message for streaming.